dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/ProgramController.java

```diff
@@ -125,6 +138,28 @@ public ResponseEntity<MetadataExportParams> getProgramWithDependencies( @PathVar
         return ResponseEntity.ok( exportParams );
     }
 
+    @PostMapping( "/{uid}/copy" )
+    @ResponseStatus( HttpStatus.CREATED )
+    @ResponseBody
+    public WebMessage copyProgram( @PathVariable( "uid" ) String uid,
+        @RequestParam( required = false ) Map<String, String> copyOptions )
+        throws NotFoundException,
+        ConflictException
+    {
+        Program programCopy = null;
+        try
+        {
+            programCopy = copyService.copyProgramFromUid( uid, copyOptions );
+        }
+        catch ( DataIntegrityViolationException dive )
+        {
+            throw new ConflictException(
+                Objects.requireNonNull( dive.getRootCause() ).getMessage() );
```

Comment:
With the use of 'requireNonNull' here it provides a nicer, more detailed message if an error occurs. e.g. when the Program name is too long during a copy, the error message would look like this `value too long for type character varying(230)`
Without the 'requireNonNull', we can't get the root cause and the error would look like this e.g. `could not execute statement; SQL [n/a]; nested exception is org.hibernate.exception.DataException: could not execute statement`

Reply:
the `dive` exception variable can throw a NPE from the call to `getRootCause()`. Sonar will not allow a call to `getRootCause()` without the `requireNonNull`.
So we can either keep as is with the more detailed message or remove the `requireNonNull` and have the message which does not give any idea why the error occurred.

Reply:
In other words the `getRootCause()` can return `null`? We should not express the expectation it is never null when it can be. That is confusing. If it can be null just assign it to a local variable and check if it is and if so generate a message from the `dive` exception, e.g. use its `getMessage()` in case of `null`.